draft-west-cookie-incrementalism.md

```diff
@@ -401,85 +404,303 @@ from "have the same registrable domain" (or "is an exact match for") to say
+    which sets the same cookie. This second cookie would match the
+    first and then overwrite it. 
+2. https://example.com:123 sets a cookie `Set-Cookie: a=domaincookie;
+    Domain=example.com`. The user then visits https://example.com:456
+    which sets `Set-Cookie: a=origincookie`. The second cookie would
+    match the first and overwrite it.
 
-First, alter the Storage Model defined in Section 5.4 of {{RFC6265bis}} by adding `scheme` to the
-list of fields the user agent stores about each cookie, and setting it when creating a cookie by
-altering step 2 of the same algorithm from:
+Given that these “domain cookies'' are less trusted (due to their
+accessibility by different hosts and ports) we should try and prevent
+them from shadowing “origin cookies”. Example:
+
+1. https://example.com:456 sets a cookie `Set-Cookie: a=origincookie`.
+   The user then visits https://example.com:123 which tries to set a
+   cookie `Set-Cookie: a=domaincookie; Domain=example.com`. The second
+   cookie would fail to be set because the cookie it’s attempting to
+   overwrite is an origin cookie.
+
+Finally, we don’t ever want to allow a cookie to pass between schemes,
+given the huge security differences between them. So there should be no
+way for a server to specify that a given cookie should be sent to a
+different scheme.
+
+How we accomplish this is as follows:
+
+First, add the concept of port matching which helps to simplify checking
+if a cookie would match a port value. We can do that by adding a new
+section under 5.1 (this new section depends on the modification to 5.4
+below)
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time.
+5.1.5 Port matching 
+
+An integer port-matches a given cookie if any of the following
+conditions are true:
+  1. The integer exactly matches the cookie’s port value.
+
+  2. The cookie’s host-only-flag is false.
 ~~~
 
-to:
+We also want to add a sub-algorithm to protect origin cookies from being
+overwritten/shadowed by domain cookies. This’ll be run during the
+creation of new cookies in a later step.
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time. Set the scheme to request-uri's origin's scheme
-     component.
+5.1.6 New Cookie does not Shadow or Overwrite Origin-only Cookies
+
+This algorithm protects any existing origin-only cookies from being
+overwritten or shadowed by a new cookie ("new-cookie") with the Domain
+attribute and is as follows:
+
+If the new-cookie's host-only-flag is true, then return true.
+
+If the new-cookie’s host-only-flag is false, then return false if the
+cookie store contains at least one cookie ("equiv-cookie") that meets
+all of the following criteria: 
+  1. Equiv-cookie’s name and scheme match new-cookie’s name and scheme.
```

Comment:
nit: how about "are equal to" instead of "match" to avoid confusion with other notions of "match" defined above?